### PR TITLE
Infer needed `zero_alloc` checks from signatures

### PR DESCRIPTION
**Pull request by ccasin**
In the "`zero_alloc` in signatures" work, we made it so that to expose a function as zero alloc in your signature, you must also put a `zero_alloc` attribute on it in the structure.  This guarantees the backend check will run on that function, as implied by the signature.  But also it's a real pain to write the attribute in two places, and we can do better: This PR lets the user skip writing the attribute on the implementation and infers that the check is needed from the signature.

There are three commits, the first two of which are refactoring.  I suggest the main reviewer reads commit by commit.
1) The first commit makes a change that was suggested for the original signatures work.  It changes the way we store `zero_alloc` attributes on applications to be the syntactic version of those attributes rather than their values in the semantic domain.  At the same time, it changes when we check whether we can infer `[@zero_alloc assume]` on an application - now we do this during translation to lambda rather than when typechecking the application.  This is helpful in later commits - we are delaying that inference step until it is safe to default `zero_alloc` variables.
2) The second commit moves the logic for checking whether one `zero_alloc` attribute is at least as strong as another out of `Include_core` and into its own file.  It makes sense to have this together with the implementation of `zero_alloc` variables (which comes in the next commit), and I've made it its own commit so that I'm not both moving and changing that code in one commit.
3) The third commit implements inference and adds tests. If a function definition has no `zero_alloc` attribute on it we make a variable representing whether or not it will need to be checked.  That variable can be filled in when we compare the module to its signature.  These variables are somewhat simpler than jkind or type variables because we never need to link two of them.

I suggest @ncik-roberts reviews, as he did for the signatures implementation.  @gretay-js should take a look at the tests, at least.

This PR does not yet make it so that zero_alloc information is carried across aliases (`let f = g`) - there is enough here that I think that should be its own PR.

One minor design note:

### Locations in errors

Today `zero_alloc` errors look like this:

```
File "/home/ccasinghino/tmp/foo.ml", line 1, characters 5-15:
1 | let[@zero_alloc] f x = (x,x)
         ^^^^^^^^^^
Error: Annotation check for zero_alloc failed on function Foo.f (camlFoo__f_0_1_code)

File "/home/ccasinghino/tmp/foo.ml", line 1, characters 23-28:
1 | let[@zero_alloc] f x = (x,x)
                           ^^^^^
Error: allocation of 24 bytes
```

The error has two parts: The first identifies the location of the attribute that said to check something.  The second shows the offending allocation.

After this PR, the attribute that caused the check may be far removed from the implementation, off in a signature somewhere.  Making that show up first in the error seems bad for editor integration - people probably want to navigate to the offending function.  So if the check is an inferred one, I've made the first error point to the function that was checked instead.

I'm not sure that's exactly the right place to point.  Also it would probably be convenient to know what attribute caused this check, somehow, even if that's not going to be the place we most commonly want to navigate.  That information is easy to make available, but I think the right thing is to pass it to the backend and have the backend include it later in the error, and propose putting that off for a future PR.

### CODE REVIEW

File: tests/backend/zero_alloc_checker/test_inference.ml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+   [val[@zero_alloc] f_arity_two : int -> int -> int] we'll get a type error.
+   Here we show that if we give it a signature that passes the typechecker (the
+   mli has [val[@zero_alloc (arity 1)] f_arity_two : int -> int -> int]), it
+   correctly gets checked and gives a zero_alloc backend error. *)
+let f_arity_two x = fun y -> x + y
+
+
+(* Shadowing: the mli marks [f_shadow] zero_alloc.  That check should only apply
+   to the last such function in the file, so we get no error even though there
+   is an earlier function with the same name and type that allocates. *)
```
Comment (ncik-roberts) [inline tests/backend/zero_alloc_checker/test_inference.ml:32]: Can you please write a test in the other direction, too? (You get an error if the shadowing function allocates, and the mli is marked zero-alloc?)

File: tests/backend/zero_alloc_checker/test_inference.ml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  if x = 42 then () else
+    let s = Printf.sprintf "%d\n" x in
+    raise (E s)
+
+(* These are tagged [@zero_alloc opt] in the mli - the latter should fail just
+   when the flag to check opt things is passed. *)
+let f_opt_success x =
+  if x = 42 then () else
+    let s = Printf.sprintf "%d\n" x in
+    raise (E s)
```
Comment (ncik-roberts) [inline tests/backend/zero_alloc_checker/test_inference.ml:62]: Is it easy to write a test of a function that is marked as `[@zero_alloc opt]` in the mli, and that _would_ fail if it was marked as `[@zero_alloc]`? (I.e. a test where the higher level of optimization removes allocation.)

**Review by ncik-roberts — COMMENTED** [review]
Just a review on tests so far.

File: tests/backend/zero_alloc_checker/test_inference.ml (hunk 2 above)
Comment (ccasin) [reply tests/backend/zero_alloc_checker/test_inference.ml:62]: I think no: For users we describe `opt` things as being checked when compiling with optimization, but actually it's just about whether you pass a flag saying the opt things should be checked (which people tell the build system to do when building with those higher optimization levels).

**Comment by ccasin** [comment]
I've reworked this slightly to offer nicer inference behavior, and rebased/squashed to give it a nice history.  Nothing about the original PR description needs changing.

One interesting note, pointed out by Leo, is that the rules for zero_alloc variables are similar to the rules for modalities on the comonadic (prescriptive) axes.  In typemod I've slightly generalized the infrastructure to support those that was just added in #2685 and renamed the functions to, hopefully, explain their purpose more clearly, make clear that they are doing something similar to `remove_mode_and_kind_variables`, and be clear they deal with both modalities and zero_alloc variables.

File: ocaml/testsuite/tests/typing-zero-alloc/signatures.ml [hunk 3]
```diff
@@ -1002,3 +936,257 @@ module type S_no_assume = sig val f : int -> int * int [@@zero_alloc] end
...
+(* When you take the module type of a module, you get the current state of its
+   variables.  The computed signature itself has no variables, and the original
+   module keeps its variables (but changing them won't change the signature you
+   got the first time, just the signature you'll get if you do [module type of]
+   again). *)
+module M_for_mto = struct
+  let f x = x+1
+end
+
+(* The current state of the variales is "no check" *)
```
Comment (ncik-roberts) [inline ocaml/testsuite/tests/typing-zero-alloc/signatures.ml:1140]:
```suggestion
(* The current state of the variables is "no check" *)
```

File: tests/backend/zero_alloc_checker/test_inference.mli [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+val[@zero_alloc] f_basic_fail : 'a -> 'a * 'a
+
+val[@zero_alloc strict] f_strict_success : int -> int
+val[@zero_alloc strict] f_strict_fail : int -> unit
+
+val[@zero_alloc opt] f_opt_success : int -> unit
+val[@zero_alloc opt] f_opt_fail : int -> int * int
+
+val[@zero_alloc strict opt] f_strict_opt_success : int -> int
+val[@zero_alloc strict opt] f_strict_opt_fail : int -> unit
```
Comment (ncik-roberts) [inline tests/backend/zero_alloc_checker/test_inference.mli:25]: I think we should add a test demonstrating the "spooky action at a distance" of a module inclusion check causing the zero-alloc check to flag an issue with the original module definition. (You have type-checking tests for something like this.)

e.g.
```ocaml
module type S = functor () -> sig val[@zero_alloc] f : 'a -> 'a option end

module F () = struct
  let f x = Some x
end

module _ : S = F
```

In particular, the error message does not identify why the zero-alloc check is running at all, and I think it would be good to have a test demonstrating the iffy error message (and even better to someday make it better).

File: ocaml/testsuite/tests/typing-zero-alloc/cmi_test.ml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+         sig val f : int -> int [@@zero_alloc] end
+       is not included in
+         sig val f : int -> int [@@zero_alloc strict] end
+       Values do not match:
+         val f : int -> int [@@zero_alloc]
+       is not included in
+         val f : int -> int [@@zero_alloc strict]
+       The former provides a weaker "zero_alloc" guarantee than the latter.
+       File "cmi_test_lib.ml", line 7, characters 6-7: Actual declaration
+|}]
```
Comment (ncik-roberts) [inline ocaml/testsuite/tests/typing-zero-alloc/cmi_test.ml:56]: Consider adding a test showing that `M_constrained_variable` is contained in `val[@zero_alloc] f : int -> int`.

File: ocaml/typing/typedtree.mli [hunk 6]
```diff
@@ -296,10 +296,10 @@ and expression_desc =
...
                         ])
 
-            The [Zero_alloc_utils.Assume_info.t] records the optional
-            [@zero_alloc assume] attribute that may appear on applications.  If
-            that attribute is absent, it is [Assume_info.none].
-          *)
+            The [zero_alloc_attribute] records the optional [@zero_alloc assume]
+            attribute that may appear on applications.  Invariant: it is either
+            [Default_zero_alloc] (if the user wrote no attribute) or [Assume]
+            (if they did). *)
```
Comment (ncik-roberts) [inline ocaml/typing/typedtree.mli:302]: Why not mint a new type and remove the failure cases? (I'm currently confused how `Ignore_assert_all` is ruled out by type-checking, and I think minting a new type here makes that easier to see/understand.)

**Review by ncik-roberts — COMMENTED** [review]
some review

File: ocaml/typing/zero_alloc.mli [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+   zero-alloc-ness, and with what configuration (strict, opt, etc). It can be a
+   variable which will be filled in when the module the function is in is
+   compared against its signature, allowing to infer zero-alloc checks. *)
+type t
+
+(* [default] corresponds to [Default_zero_alloc], meaning no check will be
+   done. *)
+val default : t
+
+val create : const -> t
```
Comment (ncik-roberts) [inline ocaml/typing/zero_alloc.mli:26]: Suggestion: `create_const`.

File: ocaml/typing/typedtree.ml [hunk 8]
```diff
@@ -1068,32 +1068,34 @@ let let_bound_idents_with_modes_sorts_and_checks bindings =
...
-             if arity > 0 then
-               Check { strict = false;
-                       arity;
-                       loc = Location.none;
-                       opt = false }
+             if !Clflags.zero_alloc_check_assert_all && arity > 0 then
+               Zero_alloc.create (Check { strict = false;
+                                          arity;
+                                          loc = Location.none;
+                                          opt = false })
```
Comment (ncik-roberts) [inline ocaml/typing/typedtree.ml:1084]: If you rename `Zero_alloc.create` to `Zero_alloc.create_const` as I suggested elsewhere, you'll be prompted to change the indentation of the labels of `Check`. Let's just move the constructor to a new line to reduce this kind of sensitivity.

File: ocaml/typing/typedtree.ml [hunk 9]
```diff
@@ -1068,32 +1068,34 @@ let let_bound_idents_with_modes_sorts_and_checks bindings =
...
              else
-               Default_zero_alloc
-           | Default_zero_alloc -> Default_zero_alloc
+               fn.zero_alloc
+           | Ignore_assert_all | Check _ | Assume _ -> fn.zero_alloc
          in
          Ident.Map.add id zero_alloc checks
+         (* CR ccasinghino: we could copy the zero-allocness if the vb_expr
+            is an ident. *)
```
Comment (ncik-roberts) [inline ocaml/typing/typedtree.ml:1091]: Flagging this: not sure if you intended on addressing this before merging.

**Review by ncik-roberts — COMMENTED** [review]
Thanks for taking the time to add zero_alloc variables. The behavior seems much more normal to me know. I'll take a quick look again after you respond to my (mostly) nits.

File: ocaml/typing/includecore.ml [hunk 10]
```diff
@@ -167,7 +98,9 @@ let value_descriptions ~loc env name
     loc
     vd1.val_attributes vd2.val_attributes
     name;
-  zero_alloc vd1.val_zero_alloc vd2.val_zero_alloc;
+  begin try Zero_alloc.sub_exn vd1.val_zero_alloc vd2.val_zero_alloc with
```
[the remark this answers is not in the document]
Comment (ccasin) [reply ocaml/typing/includecore.ml:101]: I think this is fairly idiomatic for the compiler (e.g., this isn't the only try/with in this function). `sub_exn` is a series of checks any of which may fail independently, and I wanted to avoid writing a bunch of gross nested matches.  I could define a letop for bind in the result monad, or something, but in the typechecker the usual style is just to raise.

(I don't feel particularly strongly though - feel free to push back more)

File: ocaml/typing/typedtree.ml (hunk 9 above)
Comment (ccasin) [reply ocaml/typing/typedtree.ml:1091]: Sorry - this was a note to myself for the next PR.  I've made the comment clearer.

**Comment by ccasin** [comment]
Thanks for the review.  I think I've addressed all the minor stuff and will play around with the types in typedtree sometime in the next few days.

File: ocaml/typing/includecore.ml (hunk 10 above)
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/typing/includecore.ml:101]: I thought all error creation appeared in tail position, except for one `if`-`then` needing to grow an `else`. If that’s true, let’s change it to return an error.

[the remark this answers is not in the document]
Comment (ccasin) [reply ocaml/typing/includecore.ml:101]: Probably I'm missing some nice way to refactor this, but it's [these](https://github.com/ccasin/flambda-backend/blob/ccd41d34bd07feff3947fbaf2f96af0756789aac/ocaml/typing/zero_alloc.ml#L112) [two](https://github.com/ccasin/flambda-backend/blob/ccd41d34bd07feff3947fbaf2f96af0756789aac/ocaml/typing/zero_alloc.ml#L118) raises that seemed annoying.

[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/typing/includecore.ml:101]: I missed the second one. Fine, then we should move the try-with into the module.

[the remark this answers is not in the document]
Comment (ccasin) [reply ocaml/typing/includecore.ml]: Good idea, done.

File: ocaml/typing/typedtree.mli (hunk 6 above)
Comment (ccasin) [reply ocaml/typing/typedtree.mli]: I've taken a stab at this. It didn't result in any major simplfications, but is probably a bit clearer - let me know what you think.

**Comment by ccasin** [comment]
> Thanks for the changes. I think the type refactoring you did in the latest commit is nice and helps make some non-obvious (or comment-enforced) invariants much clearer. I haven't checked whether additional type refactors would be nice in the same way, but you can make the call.

Thanks.  I did look for additional opportunities to make this kind of change, but didn't find them, because the other places where zero_alloc stuff appears in types or the typedtree need the option to be variables.

**Review by gretay-js — APPROVED** [review]
